File: frontend/js/views/equivalent_annual_cost.js
// Equivalent annual cost (EAC) — level annual cost over an asset's life, via
// /calc/equivalent-annual-cost. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }));
const num = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 4 }));

export async function renderEquivalentAnnualCost(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.eac.h1.title">// EQUIVALENT ANNUAL COST</span></h1>
        <p class="muted small" data-i18n="view.eac.hint.intro">
            EAC is the level yearly cost with the same present value as buying and running an asset over
            its life — the right way to compare assets with different lifespans. Annualize each option
            and pick the lower. It splits into the annual operating cost plus the capital recovery cost.
            Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.eac.h2.inputs">The asset</h2>
            <form id="eac-form" class="inline-form">
                <label><span data-i18n="view.eac.label.initial">Initial cost ($)</span>
                    <input type="number" step="1000" min="0" name="initial_cost_usd" value="100000" required></label>
                <label><span data-i18n="view.eac.label.op">Annual operating cost ($)</span>
                    <input type="number" step="500" min="0" name="annual_operating_cost_usd" value="10000"></label>
                <label><span data-i18n="view.eac.label.salvage">Salvage value ($)</span>
                    <input type="number" step="500" min="0" name="salvage_value_usd" value="20000"></label>
                <label><span data-i18n="view.eac.label.rate">Discount rate (%)</span>
                    <input type="number" step="0.1" min="0" name="discount_rate_pct" value="8" required></label>
                <label><span data-i18n="view.eac.label.years">Life (years)</span>
                    <input type="number" step="1" min="1" name="years" value="5" required></label>
            </form>
        </div>
        <div id="eac-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#eac-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            initial_cost_usd: Number(fd.get('initial_cost_usd')) || 0,
            annual_operating_cost_usd: Number(fd.get('annual_operating_cost_usd')) || 0,
            salvage_value_usd: Number(fd.get('salvage_value_usd')) || 0,
            discount_rate_pct: Number(fd.get('discount_rate_pct')) || 0,
            years: Number(fd.get('years')) || 0,
        };
        try {
            const r = await api.calcEquivalentAnnualCost(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.eac.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#eac-result');
    const eacCell = r.equivalent_annual_cost_usd != null ? money(r.equivalent_annual_cost_usd) : t('view.eac.undefined');
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.eac.h2.result">The annual cost</h2>
            <div class="cards">
                <div class="card neg"><div class="label" data-i18n="view.eac.card.eac">Equivalent annual cost</div>
                    <div class="value neg">${eacCell}</div></div>
                <div class="card"><div class="label" data-i18n="view.eac.card.caprecovery">Capital recovery</div>
                    <div class="value">${money(r.capital_recovery_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.eac.card.pvcosts">PV of costs</div>
                    <div class="value">${money(r.pv_of_costs_usd)}</div></div>
            </div>
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.eac.row.af">Annuity factor</td><td>${num(r.annuity_factor)}</td></tr>
                    <tr><td data-i18n="view.eac.row.salvagepv">Salvage present value</td><td>${money(r.salvage_pv_usd)}</td></tr>
                    <tr><td data-i18n="view.eac.row.pvcosts">PV of all costs</td><td>${money(r.pv_of_costs_usd)}</td></tr>
                    <tr><td data-i18n="view.eac.row.caprecovery">Capital recovery (annualized)</td><td>${money(r.capital_recovery_usd)}</td></tr>
                    <tr class="emph neg"><td data-i18n="view.eac.row.eac">Equivalent annual cost</td><td>${eacCell}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/equivalent_annual_cost.rs
//! Equivalent annual cost (EAC) — the level yearly cost that has the same present
//! value as owning and running an asset over its life. It is the right way to
//! compare assets with different lifespans (a cheap machine replaced every 3
//! years vs a durable one replaced every 7): annualize each and pick the lower.
//!
//! ```text
//! annuity factor    = (1 − (1+r)^−n) / r
//! PV of costs        = initial + annual operating × AF − salvage·(1+r)^−n
//! EAC                = PV of costs / AF
//!                    = annual operating + (initial − salvage PV) / AF
//! ```
//!
//! The second term is the capital recovery cost — the annualized net cost of
//! buying the asset and recovering its salvage. Distinct from `npv-irr` (project
//! NPV/IRR) and `mirr` (a return measure).

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct EacInput {
    /// Up-front purchase cost (a positive number).
    pub initial_cost_usd: f64,
    /// Recurring operating cost each year.
    #[serde(default)]
    pub annual_operating_cost_usd: f64,
    /// Salvage value received at the end of the life.
    #[serde(default)]
    pub salvage_value_usd: f64,
    pub discount_rate_pct: f64,
    pub years: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct EacResult {
    /// (1 − (1+r)^−n) / r.
    pub annuity_factor: f64,
    /// Present value of the salvage received at the end.
    pub salvage_pv_usd: f64,
    /// Net present value of every cost over the life.
    pub pv_of_costs_usd: f64,
    /// Annualized net capital cost: (initial − salvage PV) / AF.
    pub capital_recovery_usd: f64,
    /// The level equivalent annual cost.
    pub equivalent_annual_cost_usd: Option<f64>,
}

pub fn analyze(input: &EacInput) -> EacResult {
    let r = input.discount_rate_pct / 100.0;
    let n = input.years;

    let af = if n <= 0.0 {
        0.0
    } else if r == 0.0 {
        n
    } else {
        (1.0 - (1.0 + r).powf(-n)) / r
    };

    let discount = if n <= 0.0 {
        1.0
    } else {
        (1.0 + r).powf(-n)
    };
    let salvage_pv = input.salvage_value_usd * discount;

    let pv_costs = input.initial_cost_usd + input.annual_operating_cost_usd * af - salvage_pv;
    let capital_recovery = if af > 0.0 {
        (input.initial_cost_usd - salvage_pv) / af
    } else {
        0.0
    };
    let eac = if af > 0.0 {
        Some(pv_costs / af)
    } else {
        None
    };

    EacResult {
        annuity_factor: af,
        salvage_pv_usd: salvage_pv,
        pv_of_costs_usd: pv_costs,
        capital_recovery_usd: capital_recovery,
        equivalent_annual_cost_usd: eac,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.5
    }

    fn base() -> EacInput {
        EacInput {
            initial_cost_usd: 100_000.0,
            annual_operating_cost_usd: 10_000.0,
            salvage_value_usd: 20_000.0,
            discount_rate_pct: 8.0,
            years: 5.0,
        }
    }

    #[test]
    fn annuity_factor() {
        // (1 − 1.08^−5) / 0.08 = 3.992710.
        assert!((analyze(&base()).annuity_factor - 3.992710).abs() < 1e-4);
    }

    #[test]
    fn salvage_present_value() {
        // 20,000 × 1.08^−5 = 13,611.66.
        assert!(close(analyze(&base()).salvage_pv_usd, 13_611.66));
    }

    #[test]
    fn pv_of_costs() {
        // 100,000 + 10,000 × 3.992710 − 13,611.66 = 126,315.44.
        assert!(close(analyze(&base()).pv_of_costs_usd, 126_315.44));
    }

    #[test]
    fn capital_recovery() {
        // (100,000 − 13,611.66) / 3.992710 = 21,636.52.
        assert!(close(analyze(&base()).capital_recovery_usd, 21_636.52));
    }

    #[test]
    fn equivalent_annual_cost() {
        // 126,315.44 / 3.992710 = 31,636.52.
        assert!(close(analyze(&base()).equivalent_annual_cost_usd.unwrap(), 31_636.52));
    }

    #[test]
    fn eac_identity() {
        // EAC = annual operating + capital recovery.
        let r = analyze(&base());
        assert!(close(
            r.equivalent_annual_cost_usd.unwrap(),
            10_000.0 + r.capital_recovery_usd
        ));
    }

    #[test]
    fn zero_rate_simple_average() {
        let r = analyze(&EacInput {
            discount_rate_pct: 0.0,
            ..base()
        });
        // AF = 5; PV costs = 100,000 + 50,000 − 20,000 = 130,000; EAC = 26,000.
        assert!(close(r.annuity_factor, 5.0));
        assert!(close(r.equivalent_annual_cost_usd.unwrap(), 26_000.0));
    }

    #[test]
    fn no_salvage() {
        let r = analyze(&EacInput {
            salvage_value_usd: 0.0,
            ..base()
        });
        assert!(close(r.salvage_pv_usd, 0.0));
        // EAC = 10,000 + 100,000 / 3.992710 = 35,045.65.
        assert!(close(r.equivalent_annual_cost_usd.unwrap(), 35_045.65));
    }

    #[test]
    fn longer_life_lowers_capital_recovery() {
        let short = analyze(&base());
        let long = analyze(&EacInput {
            years: 10.0,
            ..base()
        });
        // Spreading the purchase over more years lowers the annual capital cost.
        assert!(long.capital_recovery_usd < short.capital_recovery_usd);
    }

    #[test]
    fn zero_years_no_eac() {
        let r = analyze(&EacInput {
            years: 0.0,
            ..base()
        });
        assert!(r.equivalent_annual_cost_usd.is_none());
    }
}
